### PR TITLE
Add compiler folder feature

### PR DESCRIPTION
**Pull request by Trottoir**
Allows to compile all the contracts under : 

`assembly/contracts`

### CODE REVIEW

**Comment by peterjah** [comment]
I understand your need, and i am sure its working, but i don't think is should be done here. Toolkit should be kept a simple as possible.
maybe this code should be put in the repository you are working on ?

**Comment by Trottoir** [comment]
> I understand your need, and i am sure its working, but i don't think is should be done here. Toolkit should be kept a simple as possible. maybe this code should be put in the repository you are working on ?

I totally agree with you, we kept this in the toolkit in order to ease the demo of tomorrow.
I can see it on the massa-as-sdk for instance

**Comment by Trottoir** [comment]
> > > I understand your need, and i am sure its working, but i don't think is should be done here. Toolkit should be kept a simple as possible. maybe this code should be put in the repository you are working on ?
> > 
> > 
> > I totally agree with you, we kept this in the toolkit in order to ease the demo of tomorrow. I can see it on the massa-as-sdk for instance
> 
> If its only for the toworow demo, maybe you can use a specific branch of the toolkit ?

Where do you suggest to put it ? @peterjah

File: commands/init/compiler/compiler.ts [hunk 1]
```diff
@@ -4,7 +4,7 @@ import { readdir, readFileSync } from "fs";
 const dirToCompile = "assembly/contracts";
 
 export async function compileAS() {
-    readdir(dirToCompile, function (err: NodeJS.ErrnoException | null, files: string[]) {
+    readDir(dirToCompile, function (err: NodeJS.ErrnoException | null, files: string[]) {
```
Comment (Trottoir) [inline commands/init/compiler/compiler.ts:7]: This is not working uWu

File: commands/init/compiler/compiler.ts [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    return fileOrdered;
+}
+
+function prepareCommand(filesOrdered: string[]) {
+    let command = "";
+    filesOrdered.forEach((file) => {
+        const contractName = file.substring(0, file.length - 3);
+        command += `npx asc ${dirToCompile}/${contractName}.ts --target release -o build/${contractName}.wasm && `;
+    });
+    return command.substring(0, command.length - 4);
```
Comment (peterjah) [inline commands/init/compiler/compiler.ts:56]: i propose:
```
    return filesOrdered
        .map((file) => `npx asc ${dirToCompile}/${file} -o build/${file.replace(".ts", ".wasm")}`)
        .join(" && ");

**Review by peterjah — CHANGES_REQUESTED** [review]

File: commands/init/compiler/compiler.ts [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+        const contractASContent = readFileSync(`${dirToCompile}/${contract}`, "utf-8");
+        contractASContent.includes("fileToBase64(") ? deployers.push(contract) : fileOrdered.push(contract);
+    });
+
+    deployers.forEach((deployer) => {
+        fileOrdered.push(deployer);
+    });
+
+    return fileOrdered;
+}
```
Comment (Thykof) [inline commands/init/compiler/compiler.ts:48]:
```suggestion
function orderForCompilation(files: string[]): string[] {
    return files.sort(contract => {
        return readFileSync(join(dirToCompile, contract), "utf-8").includes("fileToBase64(") ? 1 : -1
    });
}
```

File: commands/init/compiler/compiler.ts [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+function prepareCommand(filesOrdered: string[]) {
```
Comment (Thykof) [inline commands/init/compiler/compiler.ts:50]:
```suggestion
function prepareCommand(filesOrdered: string[]): string {
```

File: commands/init/compiler/compiler.ts [hunk 5]
```diff
@@ -0,0 +1,59 @@
+import { exec } from "child_process";
+import { readdir, readFileSync } from "fs";
+
```
Comment (Thykof) [inline commands/init/compiler/compiler.ts:3]:
```suggestion
import { join } from "path";

```

File: commands/init/compiler/compiler.ts [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+        }
+
+        const filesFiltered = files.filter((file) => file.includes(".ts"));
+
+        const filesOrdered = orderForCompilation(filesFiltered);
+        console.log(`${filesOrdered.length} files to compile`);
+
+        const command = prepareCommand(filesOrdered);
+
+        exec(command, (error, stdout, stderr) => {
```
Comment (Thykof) [inline commands/init/compiler/compiler.ts:19]:
```suggestion
        exec(command, (error, _, stderr) => {
```

**Review by Thykof — CHANGES_REQUESTED** [review]

File: commands/init/compiler/compiler.ts (hunk 2 above)
Comment (Thykof) [reply commands/init/compiler/compiler.ts:56]: and use path.join

```suggestion
    return filesOrdered
        .map((file) => `npx asc ${join(dirToCompile,file)} -o build/${file.replace(".ts", ".wasm")}`)
        .join(" && ");
```

File: commands/init/compiler/compiler.ts [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+    files.forEach(async (contract) => {
+        const contractASContent = readFileSync(`${dirToCompile}/${contract}`, "utf-8");
+        contractASContent.includes("fileToBase64(") ? deployers.push(contract) : fileOrdered.push(contract);
+    });
+
+    deployers.forEach((deployer) => {
+        fileOrdered.push(deployer);
+    });
+
+    return fileOrdered;
```
Comment (peterjah) [inline commands/init/compiler/compiler.ts:47]: i propose:
```
    return files.sort((a, b) => {
        const aContent = readFileSync(`${dirToCompile}/${a}`);
        if (!aContent.includes("fileToBase64(")) {
            return -1;
        }
        return 1;
    });

**Review by peterjah — CHANGES_REQUESTED** [review]

**Review by Thykof — APPROVED** [review]

File: commands/init/compiler/compiler.ts [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+            console.log("Compilation succeeds");
+            return;
+        });
+    });
+}
+
+// Order files retrieved in the /contracts folder
+function orderForCompilation(files: string[]) {
+    return files.sort(contract => {
+        return readFileSync(join(dirToCompile, contract), "utf-8").includes("fileToBase64(") ? 1 : -1;
```
Comment (gregLibert) [inline commands/init/compiler/compiler.ts:37]: This magic should be documented

File: commands/init/compiler/compiler.ts [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+        const filesFiltered = files.filter((file) => file.includes(".ts"));
+
+        const filesOrdered = orderForCompilation(filesFiltered);
+        console.log(`${filesOrdered.length} files to compile`);
+
+        const command = prepareCommand(filesOrdered);
+
+        exec(command, (error, _, stderr) => {
+            if (error) {
+                console.log(error);
```
Comment (gregLibert) [inline commands/init/compiler/compiler.ts:21]: Are we sure that stderr will not be needed in this case ?

File: commands/init/compiler/compiler.ts (hunk 8 above)
Comment (gregLibert) [reply commands/init/compiler/compiler.ts:37]: And we don't use file2B64 anymore.

File: commands/init/compiler/compiler.ts [hunk 10]
```diff
@@ -0,0 +1,47 @@
+import { exec } from "child_process";
+import { readdir, readFileSync } from "fs";
+import { join } from "path";
+const dirToCompile = "assembly/contracts";
+
+export async function compileAS() {
+    readdir(dirToCompile, function (err: NodeJS.ErrnoException | null, files: string[]) {
+        if (err) {
+            return console.log("Unable to scan directory: " + err);
```
Comment (gregLibert) [inline commands/init/compiler/compiler.ts:9]:
```suggestion
            return console.log("Unable to read directory: " + err);
```

**Review by gregLibert — CHANGES_REQUESTED** [review]

File: commands/init/compiler/compiler.ts [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+            if (error) {
+                console.log(error);
+                return;
+            }
+            if (stderr) {
+                console.log(stderr);
+                return;
+            }
+            console.log("Compilation succeeds");
+            return;
```
[the remark this answers is not in the document]
Comment (Thykof) [reply commands/init/compiler/compiler.ts:29]: no

File: commands/init/compiler/compiler.ts [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+        if (err) {
+            return console.log("Unable to scan directory: " + err);
+        }
+
+        const filesFiltered = files.filter((file) => file.includes(".ts"));
+
+        const filesOrdered = orderForCompilation(filesFiltered);
+        console.log(`${filesOrdered.length} files to compile`);
+
+        const command = prepareCommand(filesOrdered);
```
[the remark this answers is not in the document]
Comment (Thykof) [reply commands/init/compiler/compiler.ts]: I now use asc, not exec

**Review by gregLibert — APPROVED** [review]
:clap: code is cleaner.

**Comment by gregLibert** [comment]
@peterjah waiting for your feedbacks before merging.

**Comment by peterjah** [comment]
> @peterjah waiting for your feedbacks before merging.

I already expressed my point of view on this.
To summarize, i think this is a specific need and it should not be part of the toolkit.
Toolkit init is a simple projet used for smart contract builder beginning with assembly script and massa.
We should keep it as simple as possible.
Moreover i think we spend time on rewritting some code which is already present here: https://github.com/massalabs/massa-sc-utils/tree/master/src/utils

we should factorize this.

**Comment by peterjah** [comment]
anyway i m not against merging it. but we have to think to factorize code, and i think deployer functions should be put elsewhere!

**Review by peterjah — APPROVED** [review]

**Pull request merged**